### DISCUSSION TITLE
Change a relations manager on the collection level

### DISCUSSION THREAD

**Discussion by StaffCollab** [filamentphp/filament#7028] [Help] [answered]
I'm having trouble with a relations manager who returns duplicated results.
It is a many-to-many relationship, and because this table has more columns and is doing multiple things, we have some duplicated rows on the relationship causing it to return duplicated results.

Now, as I asked in the past here about similar stuff, I was answered that I can customize the query using getTableQuery(), the problem is that I cannot get this to return only unique results.
I tried this
~~~
    public function getTableQuery(): Relation | Builder
    {
        return parent::getTableQuery()
            ->groupBy('account_advertisement_charges.id');
    }
~~~
But it is not working as I'm getting a duplicate column id error, but as I'm not sure how the select query works under the hood I cannot fix it.
Any help?

I wish I would be able to customize the collection after getting the table query so I can do a simple `->unique('id')` method

**Comment by StaffCollab** [comment]
And I actually noticed now a very weird thing. I added a distinct() method on the relationship like this
~~~
    public function charges(): BelongsToMany
    {
        return $this->belongsToMany(
            AccountAdvertisementCharges::class,
            'account_payment_charges',
            'payment',
            'charge',
        )->distinct('account_advertisement_charges.id');
    }
~~~
And it actually works to make the relationship return only unique results
But the very weird thing I'm now encountering is, now I will always get duplicated results in filament, but the pagination shows this `Showing 1 to 10 of 2 results`

And in the db there are 13 non-unique results, and 2 unique... weird

**Comment by StaffCollab** [comment]
![Screenshot 2023-07-17 004938](https://github.com/filamentphp/filament/assets/111918553/9de8a792-2056-48d4-af9f-c4590faf7694)

**Reply by StaffCollab** [reply]
I understand that a distinct statement can cause issues with pagination (as distinct with limit and offsets doesn't work well together), but even if I disable the pagination on that table, I'm still not getting the distinct results.
I do receive the distinct results on a regular relationship call (tried in Tinker now)

**Comment by pboivin** [comment]
Have you tried adding the `distinct()` call in `getTableQuery()` instead of on the relationship?

**Reply by StaffCollab** [reply]
Thanks for your help.

Unfurtanetly, the issue persists. Even if I put the `distinct() ` call on the query

**Comment by StaffCollab** [comment, marked as answer]
So I finally decided to drop the relationship and build my own eloquent builder like this.
~~~
    public function getTableQuery(): Relation | Builder
    {
        return AccountAdvertisementCharges::whereRelation('payments','account_payment.id', $this->ownerRecord->id);
    }
~~~
And it works perfectly.
